### PR TITLE
Refactor ExprExceptionContext to expose itself to be a resusable module

### PR DESCRIPTION
**Pull request by tanjialiang**
Fuzzer will need the persistency functionality ExprExceptionContext can provide.
Refactor the code so that it is public and could be used by other component.

### CODE REVIEW

File: velox/expression/Expr.h [hunk 1]
```diff
@@ -63,6 +63,62 @@ struct ExprStats {
...
 
+/// Data needed to generate exception context for the top-level expression.
+class ExprExceptionContext {
+ public:
+  ExprExceptionContext(
+      const Expr* FOLLY_NONNULL expr,
+      const RowVector* FOLLY_NONNULL vector)
+      : expr_(expr), vector_(vector) {}
+
+  void storeVector(
```
Comment (mbasmanova) [inline velox/expression/Expr.h:74]: Would it make sense to combine storeVector and storeSql methods into one method captureDataAndSql?

Why do we pass vector and expr to these methods? Aren't these available as member variables?

File: velox/expression/Expr.h [hunk 2]
```diff
@@ -63,6 +63,62 @@ struct ExprStats {
...
+      const char* FOLLY_NONNULL basePath,
+      const BaseVector& vector);
+
+  void storeSql(const char* FOLLY_NONNULL basePath, const Expr& expr);
+
+  const Expr* FOLLY_NONNULL expr() {
```
Comment (mbasmanova) [inline velox/expression/Expr.h:80]: Should this method be const?

File: velox/expression/Expr.h [hunk 3]
```diff
@@ -63,6 +63,62 @@ struct ExprStats {
...
+    return expr_;
+  }
+
+  const RowVector* FOLLY_NONNULL vector() const {
+    return vector_;
+  }
+
+  const std::string& serializedVectorPath() const {
```
Comment (mbasmanova) [inline velox/expression/Expr.h:88]: nit: dataPath?

File: velox/expression/Expr.h [hunk 4]
```diff
@@ -63,6 +63,62 @@ struct ExprStats {
...
+  }
+
+  const std::string& sqlPath() const {
+    return sqlPath_;
+  }
+
+ private:
+  /// Generates a file path in specified directory. Returns std::nullopt on
+  /// failure.
+  std::optional<std::string> generateFilePath(
```
Comment (mbasmanova) [inline velox/expression/Expr.h:99]: Can this be a function in the anonymous namespace in the .cpp file?

File: velox/expression/Expr.h [hunk 5]
```diff
@@ -63,6 +63,62 @@ struct ExprStats {
   }
 };
 
+/// Data needed to generate exception context for the top-level expression.
+class ExprExceptionContext {
```
Comment (mbasmanova) [inline velox/expression/Expr.h:67]: Would you document public methods of this class?

**Review by mbasmanova — COMMENTED** [review]
@tanjialiang Some questions.

**Comment by tanjialiang** [comment]
Thanks @mbasmanova , addressed comments

File: velox/expression/Expr.cpp [hunk 6]
```diff
@@ -408,20 +341,19 @@ std::string onTopLevelException(VeloxException::Type exceptionType, void* arg) {
...
   if (strlen(basePath) == 0) {
-    return context->expr->toString();
+    return context->expr()->toString();
   }
 
   // Save input vector to a file.
-  if (context->serializedVectorPath.empty()) {
-    context->serializedVectorPath = storeVector(basePath, *context->vector);
-    context->sqlPath = storeSql(basePath, *context->expr);
+  if (context->dataPath().empty()) {
```
Comment (mbasmanova) [inline velox/expression/Expr.cpp:348]: Would it make sense to move this check into persistDataAndSql method?

**Review by mbasmanova — APPROVED** [review]
@tanjialiang Looks good to me. I'm curious how do you plan to use this in the fuzzer.

**Comment by tanjialiang** [comment]
> @tanjialiang Looks good to me. I'm curious how do you plan to use this in the fuzzer.

* First step, persistency: Persist is straightforward. When a fuzzer test fails just create this ExprExceptionContext to persist.
* Second step, repro: Repro needs some additional components implemented 
   1. how to read and rebuild, I think these are already somewhere in some tests and just need to be abstracted and reused.
   2. how to hookup with cli for an easy repro.

**Comment by mbasmanova** [comment]
@tanjialiang Thank you for clarifying.

> Persist is straightforward. When a fuzzer test fails just create this ExprExceptionContext to persist.

I wonder if we could also make it so that we persist the inputs on crashes, not only failures. Maybe we could add a flag to Fuzzer for persist inputs before running an expression. Then, we could use that flag together with --seed to repro and capture inputs.

**Comment by tanjialiang** [comment]
> @tanjialiang Thank you for clarifying.
> 
> > Persist is straightforward. When a fuzzer test fails just create this ExprExceptionContext to persist.
> 
> I wonder if we could also make it so that we persist the inputs on crashes, not only failures. Maybe we could add a flag to Fuzzer for persist inputs before running an expression. Then, we could use that flag together with --seed to repro and capture inputs.

That makes sense! Thanks Masha

**Pull request closed without merging**